src/uct/ib/ud/base/ud_ep.c

```diff
@@ -1152,6 +1146,19 @@ ucs_status_t uct_ud_ep_flush(uct_ep_h ep_h, unsigned flags,
         UCT_TL_EP_STAT_FLUSH_WAIT(&ep->super);
     }
 
+    if (ep->flags & UCT_UD_EP_FLAG_FLUSH_CANCEL) {
+        ucs_assert(flags & UCT_FLUSH_FLAG_CANCEL);
+
+        /* Set `acked_psn` to the `psn` of last TX operation to make sure
+         * completing flush(CANCEL) and preceding TX operation when getting
+         * corresponding CQE */
+        ep->tx.acked_psn = (uct_ud_psn_t)(ep->tx.psn - 1);
+
+        if (status == UCS_OK) {
+            ucs_assert(ucs_queue_is_empty(&ep->tx.window));
+        }
+    }
```

Comment:
seems this is equivalent to doing window_release from uct_ud_ep_purge, except maybe the assertion that window became empty.
maybe move the assertion that window is empty outside of uct_ud_ep_purge() function, and do it whenever you assed window_release=1 to uct_ud_ep_purge().